### ISSUE TITLE
Cron should fail with an error if a course doesn't exist and retain previous data

### ISSUE THREAD

**Issue by jonespm** [tl-its-umich-edu/my-learning-analytics#292]
If there is no data for the specified course when retrieving from UDW it should print out an error to the cron logs. 

In addition the data should not be deleted and retained with the previous data.

Some data remaining is better than no data!

This does not have to send notifications or do anything other than exit the cron if it detects a potential issue.

**Comment by jonespm** [comment]
This might be a little more involved @DanielDugan but would be nice to have after the events of today. 

The crons are all run in crons.py via "do(self):"

I think the first method "update_with_udw_course" should fail if any of the SQL queries come back with no data. 

This method will need to be rewritten to not delete the records before checking. Maybe *before* this is run we can have a new method like:

verify_udw_courses_are_valid and run a select query just to verify the UDW courses all exist. It could just be like a count (*) where in (course_id) or something? 

This is going to be an involved task so let me know if you have any thoughts about it.

**Comment by pushyamig** [comment]
I think once the cron fails an email should be sent out to a particular group so that somebody can do a force cron update. On Openshift email host: mail-relay.itd.umich.edu

**Comment by jonespm** [comment]
Sounds good, this would need to be tested and the values EMAIL_HOST, EMAIL_PORT etc would have to be configurable.

https://docs.djangoproject.com/en/2.1/topics/email/

**Comment by pushyamig** [comment]
It will be good to have this issue for winter pilot courses. @DanielDugan, as discussed, let me know when you want to work on this together

**Comment by DanielDugan** [comment]
Let's plan on co-working on it tomorrow @pushyamig ? Please let me know what time works best for you. I want to focus on getting https://github.com/tl-its-umich-edu/student_explorer/issues/15 done today.

**Comment by DanielDugan** [comment]
Spoke with Pushyami about this Wednesday. I don't have enough info on the data structure to feel 100% on this fix. Would also need to include SMTP handler with logger.

From dashboard/cron.py

    # update file records from UDW
    def update_with_udw_file(self):
        # cron status
        status = ""

        logger.debug("in update with udw file")

        # if course ID's exists, delete records. Else alert and retain data
           Course_exists = f"""select concat({settings.UDW_FILE_ID_PREFIX}, canvas_id) as ID, 
           display_name as NAME, course_id as COURSE_ID from file_dim
           if len(Course_exists[2])>0:
                 status += deleteAllRecordInTable("file")
          else:
                error_mail_subject = "ERROR: Course missing %s on %s" % (Course_exists[2], 
                datetime.datetime.now())
                error_mail_handler = logging.handlers.SMTPHandler(SMTP_HOST, 'dataErrors@myLA', 
                [LOG_EMAIL], error_mail_subject)

        #select file record from UDW
        for UDW_course_id in settings.DEFAULT_UDW_COURSE_IDS:
            file_sql = f"""select concat({settings.UDW_FILE_ID_PREFIX}, canvas_id) as ID, display_name as 
           NAME, course_id as COURSE_ID from file_dim
                        where file_state ='available' 
                        and course_id='{UDW_course_id}'
                        order by canvas_id
                        """

            status += util_function(UDW_course_id, file_sql, 'file')
        return status

**Comment by jonespm** [comment]
Yeah. I'm not sure we need to worry about the notification part for now. I'd be happy if it just detects there is a problem. After #305 there's some tables that are going to be removed. So I think we should have a new method right at the beginning of this that verifies that this can work. 

Something like 
```
courses = db_util.get_supported_courses()
courses_str = ','.join(courses)

# If courses exist in the UDP which is going to be something like
            course_sql = f"select count(id) from course_dim where id in ({courses_str})'"
```

An example of what this should generate be like
`SELECT count(id) FROM course_dim WHERE id in(17700000000109008,17700000000000349);`

And this value from the db "should" return the same number of courses as are in the courses list. If it doesn't just try to do a write a message and do a `sys.exit('some error message')`as a first step.

**Comment by jonespm** [comment]
The code has changed a little since then so not sure if you had any code from back then and some of the comments on that ticket look confusing. I think it's still somewhat similar.

My idea for this was really that nothing should be deleted if there is a course that is missing that is intended to be loaded. We don't need any notifications or anything right now. I'd rather it just get the exception up to the cron.

So at the beginning of the do(self): method, it should call a method like

verify_courses_exist()

And that method will just run an pseudocode that is like
      courses_supported = <generate a CSV list of the courses supported from the Course table in the local database>
      count = executeDbQuery(f"count(id) from course_dim where id in ({courses_supported})") # You may need to check that this returns the result
      if (count = courses_supported.length)
        # Great it's a match
      else
        # No match, raise an Exception

I'll add these details to the ticket.

**Comment by DanielDugan** [comment]
Creating a pull request now.
Cron fails when a course is missing data from UDW.
Will continue to work to clarify the error messaging.

[2019-01-15 10:58:52 -0500] [89] [DEBUG] (0.001) None; args=['dashboard.DashboardCronJob', '2019-01-15 15:58:52.051426', '2019-01-15 15:58:52.946022', False, 'ser_id AS sis_id,\n                        p.unique_name AS sis_name,\n                        u.global_canvas_id AS id,\n                        c.current_score AS current_grade,\n                        c.final_score AS final_grade,\n                        \'17700000000Daniel Fails\' as course_id\n                        from user_dim u,\n                        pseudonym_dim p,\n                        course_score_fact c,\n                        (select e.user_id as user_id, e.id as enrollment_id from enrollment_dim e\n                        where e.course_id = \'17700000000Daniel Fails\'\n                        and e.type=\'StudentEnrollment\'\n                        and e.workflow_state=\'active\' ) as e\n                        where p.user_id=u.id\n                        and u.id = e.user_id\n                        and c.enrollment_id =  e.enrollment_id\n                        and p.sis_user_id is not null\n                        \': invalid input syntax for integer: "17700000000Daniel Fails"\n\n', None]

**Comment by pushyamig** [comment]
@zqian could you add test plan on how to test this? as this is for Win19 release

**Comment by pushyamig** [comment]
I am closing this issue as coding is done and closing the issue will get this  to QA column in Project board

**Comment by jennlove-um** [comment]
This works as designed.  When an invalid course ID was added to MyLA, the cron generated an error message.  No data was deleted from the tables, so MyLA still displays data from prior cron fun.
